File: app-macos/Sources/RunnerForge/Services/SelfTestService.swift
import Foundation

public struct SelfTestProgress: Sendable {
    public let stage: String
    public let detail: String
}

/// The in-app self-test: one button that proves this setup currently works.
///
/// It dispatches the self-hosted self-test workflow, waits, downloads every
/// artifact, RE-ASSERTS the artifact contract locally rather than trusting the
/// CI job's word, runs the Reaper, and writes the result to
/// `verification.lastProof`.
///
/// The check that makes it meaningful is the runner-name check: a job that
/// silently landed on a GitHub-hosted runner proves nothing about YOUR runners.
public actor SelfTestService {
    private let logBus: LogBus
    private let configStore: ConfigStore
    private let actionsService: GitHubActionsService
    private let reaperService: ReaperService
    private let sweeperService: SweeperService
    private let processRunner: ProcessRunner

    public var workflowFile = "selftest-selfhosted.yml"
    public var gitRef = "main"
    public var projectName = "Canary"
    public var timeout: Duration = .seconds(90 * 60)

    public init(
        logBus: LogBus, configStore: ConfigStore, actionsService: GitHubActionsService,
        reaperService: ReaperService, sweeperService: SweeperService, processRunner: ProcessRunner
    ) {
        self.logBus = logBus
        self.configStore = configStore
        self.actionsService = actionsService
        self.reaperService = reaperService
        self.sweeperService = sweeperService
        self.processRunner = processRunner
    }

    public func run(
        config: ForgeConfig,
        onProgress: (@Sendable (SelfTestProgress) -> Void)? = nil
    ) async -> SelfTestResult {
        let started = Date()
        var result = SelfTestResult(startedAt: started)

        let repo = config.verification.canaryRepo.contains("/")
            ? String(config.verification.canaryRepo.split(separator: "/")[1])
            : (config.github.repos.first ?? "")

        do {
            // A unique tag echoed into run-name is how the run is identified.
            // Resolving "the newest run" is race-prone and wrong under concurrency.
            let tag = "rf-selftest-\(Int(Date().timeIntervalSince1970))-\(UUID().uuidString.prefix(8))"

            onProgress?(SelfTestProgress(stage: "Dispatching", detail: "\(workflowFile) with tag \(tag)"))

            let runId = try await actionsService.dispatchWorkflow(
                config: config, repo: repo, workflowFile: workflowFile, ref: gitRef,
                inputs: ["run_name_tag": tag, "project_name": projectName])

            result.runId = runId
            onProgress?(SelfTestProgress(stage: "Waiting", detail: "run \(runId)"))

            let conclusion = try await actionsService.waitForRun(
                config: config, repo: repo, runId: runId, timeout: timeout,
                onStatus: { message in onProgress?(SelfTestProgress(stage: "Waiting", detail: message)) })

            result.conclusion = conclusion
            result.durationSeconds = Int(Date().timeIntervalSince(started))

            // --- did it land on OUR runners? -----------------------------------
            onProgress?(SelfTestProgress(stage: "Checking runners", detail: "reading job runner names"))

            let jobs = try await actionsService.listJobs(config: config, repo: repo, runId: runId)
            result.runnerNames = jobs.map { $0.runnerName ?? "unknown" }

            for job in jobs {
                await logBus.info("selftest",
                    "  \(job.name): \(job.conclusion ?? "?") on \(job.runnerName ?? "unknown")")
            }

            if !result.allJobsOnForgeRunners {
                let offenders = jobs
                    .filter { !($0.runnerName ?? "").hasPrefix("forge-") }
                    .map { "\($0.name) (\($0.runnerName ?? "unknown"))" }
                    .joined(separator: ", ")
                await logBus.error("selftest",
                    "these jobs did not run on a Runner Forge runner: \(offenders). A job that landed on a "
                    + "hosted runner proves nothing about this setup.")
            }

            // --- artifacts -----------------------------------------------------
            onProgress?(SelfTestProgress(stage: "Downloading", detail: "artifacts"))

            let artifacts = try await actionsService.listArtifacts(
                config: config, repo: repo, runId: runId)

            let proofDir = (config.paths.workDir as NSString)
                .appendingPathComponent("proof/\(runId)")
            try FileManager.default.createDirectory(atPath: proofDir, withIntermediateDirectories: true)

            for artifact in artifacts {
                onProgress?(SelfTestProgress(stage: "Downloading", detail: artifact.name))
                _ = try await actionsService.downloadArtifact(
                    config: config, repo: repo, artifact: artifact,
                    destinationDirectory: proofDir, processRunner: processRunner)
            }

            // --- re-assert the contract LOCALLY --------------------------------
            onProgress?(SelfTestProgress(stage: "Verifying", detail: "re-asserting the artifact contract"))

            let aaxAvailable = !config.paths.aaxSdkSource.isEmpty
            let contract = ArtifactExpectation.contract(for: projectName, aaxAvailable: aaxAvailable)
            var contractSatisfied = true

            for expectation in contract {
                let outcome = await Self.verifyArtifact(
                    downloadRoot: proofDir, expectation: expectation, processRunner: processRunner)

                result.artifacts.append(SelfTestArtifact(
                    name: expectation.name, bytes: outcome.bytes,
                    fileCount: outcome.fileCount, verdict: outcome.passed ? "pass" : "fail"))

                if !outcome.passed && expectation.required {
                    contractSatisfied = false
                    await logBus.error("selftest", "artifact contract violated: \(expectation.name)")
                }
            }

            // --- reaper --------------------------------------------------------
            onProgress?(SelfTestProgress(stage: "Reaping", detail: "verifying nothing is still running"))
            let reaper = try await reaperService.reap(config: config)
            result.reaperExitCode = reaper.rawValue

            // --- sweeper -------------------------------------------------------
            onProgress?(SelfTestProgress(stage: "Sweeping", detail: "reclaiming ephemeral disk"))
            result.sweeperReclaimedBytes = Int(try await sweeperService.purge(config: config))

            // --- verdict -------------------------------------------------------
            let passed = conclusion == "success"
                && result.allJobsOnForgeRunners
                && contractSatisfied
                && reaper == .clean

            result.verdict = passed ? "pass" : "fail"
            result.durationSeconds = Int(Date().timeIntervalSince(started))

            await logBus.info("selftest", passed
                ? "SELF-TEST PASSED in \(result.durationSeconds)s (run \(runId))"
                : "SELF-TEST FAILED (run \(runId)): conclusion=\(conclusion), "
                  + "onForgeRunners=\(result.allJobsOnForgeRunners), contract=\(contractSatisfied), "
                  + "reaper=\(reaper)")
        } catch {
            result.verdict = "fail"
            await logBus.error("selftest", String(describing: error))
        }

        // Persisted so the Runners page can show, at a glance, whether this setup
        // is currently known-good.
        var updated = config
        updated.verification.lastProof = result
        try? await configStore.save(updated)

        return result
    }

    public struct ArtifactOutcome: Sendable {
        public let passed: Bool
        public let bytes: Int
        public let fileCount: Int
    }

    /// The in-app twin of scripts/verify-artifacts.sh. Untars macOS bundles
    /// before asserting, because upload-artifact flattens symlinks and the
    /// bundles are tarred on the way in specifically to survive that.
    public static func verifyArtifact(
        downloadRoot: String, expectation: ArtifactExpectation, processRunner: ProcessRunner
    ) async -> ArtifactOutcome {
        let directory = (downloadRoot as NSString).appendingPathComponent(expectation.name)

        guard FileManager.default.fileExists(atPath: directory) else {
            return ArtifactOutcome(passed: !expectation.required, bytes: 0, fileCount: 0)
        }

        let files = allFiles(under: directory)
        let bytes = files.reduce(0) { total, path in
            let attributes = try? FileManager.default.attributesOfItem(atPath: path)
            return total + ((attributes?[.size] as? NSNumber)?.intValue ?? 0)
        }

        guard bytes > 0 else {
            return ArtifactOutcome(passed: false, bytes: 0, fileCount: files.count)
        }

        var searchRoot = directory

        if expectation.untar {
            let extracted = (directory as NSString).appendingPathComponent("__untarred")
            try? FileManager.default.createDirectory(atPath: extracted, withIntermediateDirectories: true)

            let tarballs = files.filter { $0.hasSuffix(".tar") }
            guard !tarballs.isEmpty else {
                return ArtifactOutcome(passed: false, bytes: bytes, fileCount: files.count)
            }

            for tarball in tarballs {
                // -p preserves permissions and restores the symlinks that
                // upload-artifact would otherwise have flattened.
                _ = try? await processRunner.run(
                    "/usr/bin/tar", arguments: ["-xpf", tarball, "-C", extracted],
                    logSource: "selftest", streamOutput: false)
            }

            searchRoot = extracted
        }

        for glob in expectation.globs {
            let matches = entries(under: searchRoot, matching: glob)
            guard !matches.isEmpty else {
                return ArtifactOutcome(passed: false, bytes: bytes, fileCount: files.count)
            }

            let anyNonEmpty = matches.contains { match in
                var isDirectory: ObjCBool = false
                FileManager.default.fileExists(atPath: match, isDirectory: &isDirectory)
                if isDirectory.boolValue {
                    return allFiles(under: match).contains { path in
                        let attributes = try? FileManager.default.attributesOfItem(atPath: path)
                        return ((attributes?[.size] as? NSNumber)?.intValue ?? 0) > 0
                    }
                }
                let attributes = try? FileManager.default.attributesOfItem(atPath: match)
                return ((attributes?[.size] as? NSNumber)?.intValue ?? 0) > 0
            }

            guard anyNonEmpty else {
                return ArtifactOutcome(passed: false, bytes: bytes, fileCount: files.count)
            }
        }

        for bundleGlob in expectation.bundles {
            for bundle in entries(under: searchRoot, matching: bundleGlob) {
                var isDirectory: ObjCBool = false
                FileManager.default.fileExists(atPath: bundle, isDirectory: &isDirectory)
                guard isDirectory.boolValue else { continue }

                let macOsDir = (bundle as NSString).appendingPathComponent("Contents/MacOS")
                let plist = (bundle as NSString).appendingPathComponent("Contents/Info.plist")

                guard FileManager.default.fileExists(atPath: macOsDir),
                      FileManager.default.fileExists(atPath: plist) else {
                    return ArtifactOutcome(passed: false, bytes: bytes, fileCount: files.count)
                }
            }
        }

        return ArtifactOutcome(passed: true, bytes: bytes, fileCount: files.count)
    }

    static func allFiles(under directory: String) -> [String] {
        guard let enumerator = FileManager.default.enumerator(atPath: directory) else { return [] }
        var result: [String] = []

        while let relative = enumerator.nextObject() as? String {
            let full = (directory as NSString).appendingPathComponent(relative)
            var isDirectory: ObjCBool = false
            FileManager.default.fileExists(atPath: full, isDirectory: &isDirectory)
            if !isDirectory.boolValue { result.append(full) }
        }

        return result
    }

    static func entries(under directory: String, matching glob: String) -> [String] {
        guard let enumerator = FileManager.default.enumerator(atPath: directory) else { return [] }
        var result: [String] = []

        // The globs in the contract are simple "*.ext" or a literal file name.
        let suffix = glob.hasPrefix("*") ? String(glob.dropFirst()) : nil

        while let relative = enumerator.nextObject() as? String {
            let name = (relative as NSString).lastPathComponent
            let matches = suffix.map { name.hasSuffix($0) } ?? (name == glob)
            if matches { result.append((directory as NSString).appendingPathComponent(relative)) }
        }

        return result
    }
}

File: app-windows/RunnerForge/Services/SelfTestService.cs
using System.Formats.Tar;
using System.IO.Compression;
using RunnerForge.Models;

namespace RunnerForge.Services;

public sealed record SelfTestProgress(string Stage, string Detail);

/// <summary>
/// The in-app self-test: one button that proves this setup currently works.
/// </summary>
/// <remarks>
/// It dispatches the self-hosted self-test workflow, waits, downloads every
/// artifact, RE-ASSERTS the artifact contract locally rather than trusting the
/// CI job's word, runs the Reaper, and writes the result to
/// <c>verification.lastProof</c>.
///
/// The check that makes it meaningful is the runner-name check: a job that
/// silently landed on a GitHub-hosted runner proves nothing about YOUR runners.
/// </remarks>
public sealed class SelfTestService(
    LogBus logBus,
    ConfigStore configStore,
    GitHubActionsService actionsService,
    ReaperService reaperService,
    SweeperService sweeperService)
{
    private readonly LogBus _logBus = logBus;
    private readonly ConfigStore _configStore = configStore;
    private readonly GitHubActionsService _actionsService = actionsService;
    private readonly ReaperService _reaperService = reaperService;
    private readonly SweeperService _sweeperService = sweeperService;

    public string WorkflowFile { get; set; } = "selftest-selfhosted.yml";
    public string GitRef { get; set; } = "main";
    public string ProjectName { get; set; } = "Canary";
    public TimeSpan Timeout { get; set; } = TimeSpan.FromMinutes(90);

    public async Task<SelfTestResult> RunAsync(
        ForgeConfig config,
        IProgress<SelfTestProgress>? progress = null,
        CancellationToken cancellationToken = default)
    {
        DateTimeOffset started = DateTimeOffset.UtcNow;
        string repo = config.Verification.CanaryRepo.Contains('/')
            ? config.Verification.CanaryRepo.Split('/')[1]
            : config.GitHub.Repos.FirstOrDefault() ?? "";

        var result = new SelfTestResult { StartedAt = started, Conclusion = "failure", Verdict = "fail" };

        try
        {
            // A unique tag echoed into run-name is how the run is identified.
            // Resolving "the newest run" is race-prone and wrong under concurrency.
            string tag = $"rf-selftest-{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}-{Guid.NewGuid():N}"[..48];

            progress?.Report(new SelfTestProgress("Dispatching", $"{WorkflowFile} with tag {tag}"));

            long runId = await _actionsService.DispatchWorkflowAsync(
                config, repo, WorkflowFile, GitRef,
                new Dictionary<string, string> { ["run_name_tag"] = tag, ["project_name"] = ProjectName },
                cancellationToken).ConfigureAwait(false);

            result.RunId = runId;
            progress?.Report(new SelfTestProgress("Waiting", $"run {runId}"));

            var runProgress = new Progress<string>(message =>
                progress?.Report(new SelfTestProgress("Waiting", message)));

            string conclusion = await _actionsService
                .WaitForRunAsync(config, repo, runId, Timeout, runProgress, cancellationToken)
                .ConfigureAwait(false);

            result.Conclusion = conclusion;
            result.DurationSeconds = (int)(DateTimeOffset.UtcNow - started).TotalSeconds;

            // --- did it land on OUR runners? -------------------------------
            progress?.Report(new SelfTestProgress("Checking runners", "reading job runner names"));

            IReadOnlyList<WorkflowJobInfo> jobs = await _actionsService
                .ListJobsAsync(config, repo, runId, cancellationToken).ConfigureAwait(false);

            result.RunnerNames = [.. jobs.Select(j => j.RunnerName ?? "unknown")];

            foreach (WorkflowJobInfo job in jobs)
            {
                _logBus.Info("selftest", $"  {job.Name}: {job.Conclusion} on {job.RunnerName ?? "unknown"}");
            }

            if (!result.AllJobsOnForgeRunners)
            {
                string offenders = string.Join(", ",
                    jobs.Where(j => !(j.RunnerName ?? "").StartsWith("forge-", StringComparison.Ordinal))
                        .Select(j => $"{j.Name} ({j.RunnerName ?? "unknown"})"));
                _logBus.Error("selftest",
                    $"these jobs did not run on a Runner Forge runner: {offenders}. "
                    + "A job that landed on a hosted runner proves nothing about this setup.");
            }

            // --- artifacts -------------------------------------------------
            progress?.Report(new SelfTestProgress("Downloading", "artifacts"));

            IReadOnlyList<WorkflowArtifact> artifacts = await _actionsService
                .ListArtifactsAsync(config, repo, runId, cancellationToken).ConfigureAwait(false);

            string proofDir = Path.Combine(config.Paths.WorkDir, "proof", runId.ToString());
            Directory.CreateDirectory(proofDir);

            foreach (WorkflowArtifact artifact in artifacts)
            {
                progress?.Report(new SelfTestProgress("Downloading", artifact.Name));
                await _actionsService
                    .DownloadArtifactAsync(config, repo, artifact, proofDir, cancellationToken)
                    .ConfigureAwait(false);
            }

            // --- re-assert the contract LOCALLY ----------------------------
            progress?.Report(new SelfTestProgress("Verifying", "re-asserting the artifact contract"));

            bool aaxAvailable = !string.IsNullOrWhiteSpace(config.Paths.AaxSdkSource);
            IReadOnlyList<ArtifactExpectation> contract =
                ArtifactExpectation.ContractFor(ProjectName, aaxAvailable);

            bool contractSatisfied = true;

            foreach (ArtifactExpectation expectation in contract)
            {
                (bool artifactOk, long bytes, int files) = VerifyArtifact(proofDir, expectation);

                result.Artifacts.Add(new SelfTestArtifact
                {
                    Name = expectation.Name,
                    Bytes = bytes,
                    FileCount = files,
                    Verdict = artifactOk ? "pass" : "fail",
                });

                if (!artifactOk && expectation.Required)
                {
                    contractSatisfied = false;
                    _logBus.Error("selftest", $"artifact contract violated: {expectation.Name}");
                }
            }

            // --- reaper ----------------------------------------------------
            progress?.Report(new SelfTestProgress("Reaping", "verifying nothing is still running"));

            ReaperResult reaper = await _reaperService
                .ReapAsync(config, [], [], cancellationToken).ConfigureAwait(false);
            result.ReaperExitCode = (int)reaper;

            // --- sweeper ---------------------------------------------------
            progress?.Report(new SelfTestProgress("Sweeping", "reclaiming ephemeral disk"));
            result.SweeperReclaimedBytes = await _sweeperService
                .PurgeAsync(config, cancellationToken).ConfigureAwait(false);

            // --- verdict ---------------------------------------------------
            bool passed = conclusion == "success"
                          && result.AllJobsOnForgeRunners
                          && contractSatisfied
                          && reaper == ReaperResult.Clean;

            result.Verdict = passed ? "pass" : "fail";
            result.DurationSeconds = (int)(DateTimeOffset.UtcNow - started).TotalSeconds;

            _logBus.Info("selftest", passed
                ? $"SELF-TEST PASSED in {result.DurationSeconds}s (run {runId})"
                : $"SELF-TEST FAILED (run {runId}): conclusion={conclusion}, "
                  + $"onForgeRunners={result.AllJobsOnForgeRunners}, contract={contractSatisfied}, "
                  + $"reaper={reaper}");
        }
        catch (OperationCanceledException)
        {
            result.Conclusion = "cancelled";
            result.Verdict = "fail";
            _logBus.Warning("selftest", "cancelled");
        }
        catch (Exception ex)
        {
            result.Verdict = "fail";
            _logBus.Error("selftest", ex.Message);
        }

        // Persisted so the Runners page can show, at a glance, whether this
        // setup is currently known-good.
        config.Verification.LastProof = result;
        _configStore.Save(config);

        return result;
    }

    /// <summary>
    /// The in-app twin of scripts/verify-artifacts.sh. Untars macOS bundles
    /// before asserting, because upload-artifact flattens symlinks and the
    /// bundles are tarred on the way in specifically to survive that.
    /// </summary>
    public static (bool Passed, long Bytes, int FileCount) VerifyArtifact(
        string downloadRoot, ArtifactExpectation expectation)
    {
        string directory = Path.Combine(downloadRoot, expectation.Name);

        if (!Directory.Exists(directory))
        {
            return (!expectation.Required, 0, 0);
        }

        string[] files = Directory.GetFiles(directory, "*", SearchOption.AllDirectories);
        long bytes = files.Sum(f => new FileInfo(f).Length);

        if (bytes == 0) return (false, 0, files.Length);

        string searchRoot = directory;

        if (expectation.Untar)
        {
            string extracted = Path.Combine(directory, "__untarred");
            Directory.CreateDirectory(extracted);

            string[] tarballs = Directory.GetFiles(directory, "*.tar", SearchOption.AllDirectories);
            if (tarballs.Length == 0) return (false, bytes, files.Length);

            foreach (string tarball in tarballs)
            {
                try { TarFile.ExtractToDirectory(tarball, extracted, overwriteFiles: true); }
                catch (Exception) { return (false, bytes, files.Length); }
            }

            searchRoot = extracted;
        }

        foreach (string glob in expectation.Globs)
        {
            string[] matches = Directory.GetFileSystemEntries(searchRoot, glob, SearchOption.AllDirectories);
            if (matches.Length == 0) return (false, bytes, files.Length);

            bool anyNonEmpty = matches.Any(m =>
                Directory.Exists(m)
                    ? Directory.GetFiles(m, "*", SearchOption.AllDirectories).Any(f => new FileInfo(f).Length > 0)
                    : new FileInfo(m).Length > 0);

            if (!anyNonEmpty) return (false, bytes, files.Length);
        }

        foreach (string bundleGlob in expectation.Bundles)
        {
            foreach (string bundle in Directory.GetDirectories(searchRoot, bundleGlob, SearchOption.AllDirectories))
            {
                if (!Directory.Exists(Path.Combine(bundle, "Contents", "MacOS"))) return (false, bytes, files.Length);
                if (!File.Exists(Path.Combine(bundle, "Contents", "Info.plist"))) return (false, bytes, files.Length);
            }
        }

        return (true, bytes, files.Length);
    }
}
